### ISSUE TITLE
Router sometimes says method not allowed, refreshing inconsistently fixes this

### ISSUE THREAD

**Issue by ajusa** [guzba/mummy#20]
I have a small example web app called simple-twitter, which can be found here: https://github.com/ajusa/simple-twitter. I've been noticing that I will occasionally get a 405: method not allowed when browsing around on the page. That code does use https://github.com/ajusa/rowdy, but I believe the code for 405 shouldn't have anything to do with what happens in the handler - the router should be the one that is throwing this error.

I wish I had a way to consistently reproduce the issue, but I don't seem to. It's also possible that rowdy is copying something it shouldn't be? I was pretty careful to use templates and macros so that it just rewrites user code, but I could have missed something.

**Comment by guzba** [comment]
Ill have to look closer to really say, but mechanically a 405 is sent by `mummy/routers` when the incoming HTTP request method is not registered bu the URL is.  HTTP method meaning "GET" or "POST" or whatever, eg a "POST" to a registered URL, where the URL is only registered for "GET".

This may not be helpful if you already know it, but mentioning just in case to start.

**Comment by ajusa** [comment]
Yep - retrying the request sometimes fixes is, which is why I opened the issue. If it were consistent I'd be sure there was a bug in my code, but because it doesn't seem to be consistent I'm guess that the issue is localized to the mummy router.

**Comment by guzba** [comment]
I have things running and can reproduce the issue so just need a bit of time now.

**Comment by guzba** [comment]
Ok I have figured out the issue, what an adventure. This is great though, exactly what real usage reveals. I'll need to tag a release for Mummy with the fix and will update here with more info then.

**Comment by guzba** [comment]
First I got an easy repro: click "Add" to trigger a post. The following request, whether a refresh or another "Add" post will get a 405 response. Great easy repro.

Next is why? Well, after being trolled by echo not displaying \0 null bytes (which isn't really echos fault just cost me lots of time), i realized that the HTTP method looked correct when echoing but the len was wrong? Very strange.

Turns out, follow-up requests after the "Add" post / xhr request had some null bytes before the beginning of the request. This is something I will be looking more into, but once added a bit of code to notice and strip those off of the start of the incoming request data things work as expected.

This is related to keep-alive, basically one request ended and the next came in on the same TCP connection so either the bytes were sent with the previous request and I didn't remove them, or they were sent before this one, which would be strange. More to learn but the fix now is harmless.

**Comment by guzba** [comment]
Ok real real fix here is I was not reseting bytesReceived counter in the case the content length exactly matched the received bytes. Good thing to fix correctly now.

**Comment by michaelnew** [comment]
Came here to report an issue, but I just updated to 0.2.3 and it seems to be resolved (🎉). So I'm assuming it's related to this fix. But since I already typed this up I'm going to share anyway, and maybe it's helpful to know:


Sometimes but not always, mummy would crash when receiving a post request with a large body. The request headers would look like this:
`@[("Host", "137.184.6.161:42835"), ("Connection", "Keep-Alive"), ("content-length", "201587"), ("content-type", "application/json"), ("content-encoding", "gzip"), ("user-agent", "Nim httpclient/1.7.3"), ("accept-encoding", "gzip")]`

The post handler seemed to run fine, but when trying to respond it would crash with this error. `32758` seems awfully close to a signed 16 bit int.

```
index 201587 not in 0 .. 32758
fatal.nim(54)            sysFatal
mummy.nim(1414)          serve
Error: unhandled exception: fatal.nim(54)            sysFatal
index 201587 not in 0 .. 32758 [MummyError]
```

Again, I haven't been able to get this crash to happen on 0.2.3, so likely it's all better now. Also really enjoying the threaded approach to building a server, so thank you for this.
